File: StriverA2Z/Java/Arrays/Easy/UnionOfSortedArray.java
import java.util.ArrayList;

public class UnionOfSortedArray {
    public static void main(String[] args) {
        int[] arr1 = {1, 1, 2, 2, 3, 4, 5};
        int[] arr2 = {1, 2, 2, 3, 4, 4, 5, 6, 7};

        unionArray(arr1, arr2);
    }

    static void unionArray(int[] arr1, int[] arr2) {
        int i = 0;
        int j = 0;
        ArrayList<Integer> unionArr = new ArrayList<>();

        while (i < arr1.length && j < arr2.length) {
            if (arr1[i] <= arr2[j]) {
                if (unionArr.isEmpty() || unionArr.get(unionArr.size() - 1) != arr1[i]) {
                    unionArr.add(arr1[i]);
                }
                i++;
            } else {
                if (unionArr.isEmpty() || unionArr.get(unionArr.size() - 1) != arr2[j]) {
                    unionArr.add(arr2[j]);
                }
                j++;
            }
        }

        while (i < arr1.length) {
            if (unionArr.get(unionArr.size() - 1) != arr1[i]) {
                unionArr.add(arr1[i]);
            }
            i++;
        }

        while (j < arr2.length) {
            if (unionArr.get(unionArr.size() - 1) != arr2[j]) {
                unionArr.add(arr2[j]);
            }
            j++;
        }

        for (Integer integer : unionArr) {
            System.out.print(integer + " ");
        }
    }
}

File: StriverA2Z/C++/3_ArraysQuestion/3.1_Easy/union_of_sorted_array.cpp
// O(n1 + n2) both

#include<bits/stdc++.h>

using namespace std;

void unionArray(int arr1[], int n1, int arr2[], int n2) {
    int i = 0;
    int j = 0;
    vector<int> unionArr;
    while (i < n1 && j < n2) {
        if (arr1[i] <= arr2[j]) {
            if (unionArr.empty() || unionArr.back() != arr1[i]) {
                unionArr.push_back(arr1[i]);
            }
            i++;
        } else {
            if (unionArr.empty() || unionArr.back() != arr2[j]) {
                unionArr.push_back(arr2[j]);
            }
            j++;
        }
    }

    while (i < n1) {
        if (unionArr.back() != arr1[i]) {
            unionArr.push_back(arr1[i]);
        }
        i++;
    }

    while (j < n2) {
        if (unionArr.back() != arr2[j]) {
            unionArr.push_back(arr2[j]);
        }
        j++;
    }

    // printing
    for (auto it: unionArr) {
        cout << it << " ";
    }
}

int main() {
    int arr1[] = {1, 1, 2, 3, 4, 5};
    int arr2[] = {2, 3, 4, 4, 5, 6};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);
    int n2 = sizeof(arr2) / sizeof(arr2[0]);

    unionArray(arr1, n1, arr2, n2);

    return 0;
}